File: demos/react-native/modules/offline/ios/OfflineNativePlayer.swift
import AVFoundation
import ExpoModulesCore
import UIKit

final class NativePlayer: ExpoView {

    private var player: AVPlayer?
    private var loader: SasPlayerResourceLoader?
    private let loaderQueue = DispatchQueue(label: "cc.musicme.sasplayer.asset")
    private var timeObserver: Any?
    private var statusObservation: NSKeyValueObservation?
    private var rateObservation: NSKeyValueObservation?

    var autoPlay: Bool = false
    var trackTitle: String?
    var trackArtist: String?
    var trackCoverUrl: String?
    private(set) var currentRef: (cb: Int, disc: Int, track: Int)?

    let onReady = EventDispatcher()
    let onError = EventDispatcher()
    let onPlay = EventDispatcher()
    let onPause = EventDispatcher()
    let onTimeUpdate = EventDispatcher()
    let onEnded = EventDispatcher()
    let onStalled = EventDispatcher()
    let onSessionRotated = EventDispatcher()
    let onMetrics = EventDispatcher()

    private var loadStartedAt: Date?
    private var canplayAt: Date?
    private var bufferUnderruns: Int = 0
    private var lastEmittedFor: String?
    private weak var currentStreamSource: StreamSource?

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        backgroundColor = .black
    }

    deinit { NotificationCenter.default.removeObserver(self) }

    func load(cb: Int, disc: Int, track: Int) {
        guard let service = OfflineSingleton.shared.service,
              let workerUrl = PlayerConfig.shared.workerUrl,
              let tokenProvider = PlayerConfig.shared.tokenProvider
        else {
            onError(["message": "player_not_configured"])
            return
        }

        if currentRef != nil {
            emitMetrics(outcome: "aborted")
        }

        tearDownPlayer()

        loadStartedAt = Date()
        canplayAt = nil
        bufferUnderruns = 0
        lastEmittedFor = nil

        let ref = StreamSession.TrackRef(cb: cb, disc: disc, track: track)
        let trackId = "\(cb):\(disc):\(track)"
        let source: ByteSource
        if let cached = PrefetchCache.shared.take(trackId) {
            source = cached
        } else {
            do {
                source = try service.openSource(
                    ref: ref, workerUrl: workerUrl, tokenProvider: tokenProvider)
            } catch {
                onError(["message": "openSource failed: \(error)"])
                return
            }
        }
        // Evict every other cached source to keep memory bounded.
        PrefetchCache.shared.clear(except: trackId)

        self.currentStreamSource = source as? StreamSource

        let loader = SasPlayerResourceLoader(source: source)
        self.loader = loader
        let encoded = trackId.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed) ?? trackId
        guard let url = URL(string: "\(SasPlayerResourceLoader.scheme)://\(encoded)/audio.m4a") else {
            onError(["message": "invalid_trackid"]); return
        }

        let asset = AVURLAsset(url: url)
        asset.resourceLoader.setDelegate(loader, queue: loaderQueue)
        let item = AVPlayerItem(asset: asset)
        let p = AVPlayer(playerItem: item)
        self.player = p
        self.currentRef = (cb, disc, track)

        timeObserver = p.addPeriodicTimeObserver(
            forInterval: CMTime(seconds: 0.5, preferredTimescale: 1000),
            queue: .main
        ) { [weak self] time in
            guard let self = self, let d = self.player?.currentItem?.duration else { return }
            let duration = CMTimeGetSeconds(d).isFinite ? CMTimeGetSeconds(d) * 1000 : 0
            self.onTimeUpdate(["position": CMTimeGetSeconds(time) * 1000, "duration": duration])
            if let ref = self.currentRef {
                NowPlayingCenter.shared.update(
                    title: self.trackTitle ?? "\(ref.cb)/\(ref.disc)/\(ref.track)",
                    artist: self.trackArtist,
                    coverUrl: self.trackCoverUrl,
                    durationMs: duration, positionMs: CMTimeGetSeconds(time) * 1000)
            }
        }

        NotificationCenter.default.addObserver(
            self, selector: #selector(itemDidEnd),
            name: .AVPlayerItemDidPlayToEndTime, object: item)

        NotificationCenter.default.addObserver(
            self, selector: #selector(itemStalled),
            name: .AVPlayerItemPlaybackStalled, object: item)

        statusObservation = item.observe(\.status, options: [.new]) { [weak self] item, _ in
            switch item.status {
            case .readyToPlay:
                if self?.canplayAt == nil { self?.canplayAt = Date() }
                let durSec = CMTimeGetSeconds(item.duration)
                self?.onReady(["duration": (durSec.isFinite ? durSec : 0) * 1000])
            case .failed:
                self?.onError(["message": item.error?.localizedDescription ?? "unknown"])
            default: break
            }
        }
        rateObservation = p.observe(\.rate, options: [.new]) { [weak self] player, _ in
            if player.rate > 0 { self?.onPlay([:]) } else { self?.onPause([:]) }
        }

        if autoPlay { p.play() }

        NowPlayingCenter.shared.player = self
        NowPlayingCenter.shared.activate()
        NowPlayingCenter.shared.update(
            title: trackTitle ?? "\(cb)/\(disc)/\(track)",
            artist: trackArtist,
            coverUrl: trackCoverUrl,
            durationMs: 0, positionMs: 0)
    }

    @objc private func itemDidEnd() {
        emitMetrics(outcome: "canplay")
        onEnded([:])
    }

    @objc private func itemStalled() {
        bufferUnderruns += 1
        onStalled([:])
    }

    private func emitMetrics(outcome: String) {
        guard let ref = currentRef else { return }
        let trackId = "\(ref.cb):\(ref.disc):\(ref.track)"
        if lastEmittedFor == trackId { return }
        lastEmittedFor = trackId

        let firstCanplayMs: Double? = (loadStartedAt != nil && canplayAt != nil)
            ? canplayAt!.timeIntervalSince(loadStartedAt!) * 1000.0 : nil
        let totalPlayMs: Double? = loadStartedAt.map { Date().timeIntervalSince($0) * 1000.0 }
        let weakSource = currentStreamSource
        let underruns = bufferUnderruns

        Task { [weak self] in
            var bootstrap: Double? = nil
            var firstRange: Double? = nil
            var fileSize: Int64? = nil
            var rotations: Int = 0
            if let s = weakSource {
                let m = await s.sessionMetrics()
                bootstrap = m.bootstrapMs
                firstRange = m.firstRangeMs
                fileSize = m.fileSizeBytes
                rotations = m.sessionRotations
            }
            await MainActor.run {
                self?.onMetrics([
                    "v": 1,
                    "trackRef": trackId,
                    "outcome": outcome,
                    "bootstrapMs": bootstrap as Any,
                    "firstKeyMs": 0,
                    "firstRangeMs": firstRange as Any,
                    "firstCanplayMs": firstCanplayMs as Any,
                    "totalPlayMs": totalPlayMs as Any,
                    "bufferUnderruns": underruns,
                    "sessionRotations": rotations,
                    "fileSizeBytes": fileSize as Any,
                ])
            }
        }
    }

    func setPlaying(_ playing: Bool) {
        guard let p = player else { return }
        if playing { p.play() } else { p.pause() }
    }
    func seek(toMs: Double) {
        let t = CMTime(seconds: toMs / 1000.0, preferredTimescale: 1000)
        player?.seek(to: t)
    }

    private func tearDownPlayer() {
        if let ref = currentRef, lastEmittedFor != "\(ref.cb):\(ref.disc):\(ref.track)" {
            emitMetrics(outcome: "aborted")
        }
        player?.pause()
        if let t = timeObserver { player?.removeTimeObserver(t) }
        statusObservation = nil
        rateObservation = nil
        NotificationCenter.default.removeObserver(self)
        NowPlayingCenter.shared.clear()
        player = nil
        loader = nil
    }
}

/// Module-wide singleton holding the worker URL + token provider. Populated by
/// `OfflineExpoModule.configurePlayer(...)` (see Task 10).
final class PlayerConfig: @unchecked Sendable {
    static let shared = PlayerConfig()
    var workerUrl: URL?
    var tokenProvider: (@Sendable () async throws -> String)?
    var currentToken: String?
    private init() {}
}

File: demos/react-native/modules/offline/android/src/main/java/cc/musicme/offline/expo/OfflineNativePlayer.kt
package cc.musicme.offline.expo

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.annotation.OptIn
import androidx.media3.common.MediaItem
import androidx.media3.common.Player as MediaPlayer
import androidx.media3.common.util.UnstableApi
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.source.DefaultMediaSourceFactory
import cc.musicme.offline.TrackRef
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

@OptIn(UnstableApi::class)
class NativePlayer(context: Context, appContext: AppContext) : ExpoView(context, appContext) {

    val onReady by EventDispatcher()
    val onError by EventDispatcher()
    val onPlay by EventDispatcher()
    val onPause by EventDispatcher()
    val onTimeUpdate by EventDispatcher()
    val onEnded by EventDispatcher()
    val onStalled by EventDispatcher()
    val onSessionRotated by EventDispatcher()
    val onMetrics by EventDispatcher()

    var autoPlay: Boolean = false
    var trackTitle: String? = null
    var trackArtist: String? = null
    var trackCoverUrl: String? = null

    private var loadStartedAt: Long = 0L
    private var canplayAt: Long = 0L
    private var bufferUnderruns: Int = 0
    private var lastEmittedFor: String? = null
    private var currentSource: cc.musicme.offline.ByteSource? = null
    private var currentRefStr: String? = null
    private var player: ExoPlayer? = null
    private val session = MediaSessionController(context)
    private val focus = AudioFocusController(context)
    private val main = Handler(Looper.getMainLooper())
    private val tick = object : Runnable {
        override fun run() {
            val p = player ?: return
            if (p.isPlaying) {
                onTimeUpdate(mapOf(
                    "position" to p.currentPosition.toDouble(),
                    "duration" to p.duration.toDouble()
                ))
            }
            main.postDelayed(this, 500L)
        }
    }

    fun load(cb: Long, disc: Int, track: Int) {
        val svc = OfflineSingleton.service ?: run {
            onError(mapOf("message" to "player_not_configured"))
            return
        }
        val workerUrl = PlayerConfig.workerUrl
        val tokenProvider = PlayerConfig.tokenProvider
        if (workerUrl == null || tokenProvider == null) {
            onError(mapOf("message" to "player_not_configured")); return
        }
        if (currentRefStr != null) emitMetrics("aborted")
        currentRefStr = "$cb:$disc:$track"
        loadStartedAt = System.currentTimeMillis()
        canplayAt = 0L
        bufferUnderruns = 0
        lastEmittedFor = null
        player?.release()

        val trackId = "$cb:$disc:$track"
        val sourceProvider = {
            val cached = PrefetchCache.take(trackId)
            val s = cached ?: kotlinx.coroutines.runBlocking {
                svc.openSource(TrackRef(cb, disc, track), workerUrl, tokenProvider)
            }
            currentSource = s
            s
        }
        PrefetchCache.clear(except = trackId)
        val factory = SasPlayerDataSource.Factory(sourceProvider)
        val msFactory = DefaultMediaSourceFactory(context).setDataSourceFactory(factory)

        val exo = ExoPlayer.Builder(context)
            .setMediaSourceFactory(msFactory)
            .build()
        val uri = android.net.Uri.parse("${SasPlayerDataSource.SCHEME}://$cb:$disc:$track/audio.m4a")
        exo.setMediaItem(MediaItem.fromUri(uri))
        exo.addListener(object : MediaPlayer.Listener {
            override fun onPlaybackStateChanged(state: Int) {
                when (state) {
                    MediaPlayer.STATE_READY -> {
                        if (canplayAt == 0L) canplayAt = System.currentTimeMillis()
                        onReady(mapOf("duration" to exo.duration.toDouble()))
                    }
                    MediaPlayer.STATE_ENDED -> {
                        emitMetrics("canplay")
                        onEnded(emptyMap())
                    }
                    MediaPlayer.STATE_BUFFERING -> {
                        bufferUnderruns++
                        onStalled(emptyMap())
                    }
                    else -> {}
                }
            }
            override fun onIsPlayingChanged(isPlaying: Boolean) {
                if (isPlaying) onPlay(emptyMap()) else onPause(emptyMap())
            }
            override fun onPlayerError(error: androidx.media3.common.PlaybackException) {
                emitMetrics("error")
                onError(mapOf("message" to (error.message ?: "unknown")))
            }
        })
        exo.prepare()
        if (autoPlay) exo.play()
        player = exo
        session.activate(
            title = trackTitle ?: "$cb/$disc/$track",
            artist = trackArtist,
            coverUrl = trackCoverUrl,
            onPlay = { setPlaying(true) },
            onPause = { setPlaying(false) },
            onNext = { RemoteEvents.onNext?.invoke() },
            onPrev = { RemoteEvents.onPrev?.invoke() },
            onSeek = { ms -> seek(ms.toDouble()) },
        )
        focus.request(
            onLoss = { setPlaying(false) },
            onDuck = {},
            onRegain = { setPlaying(true) },
        )
        main.removeCallbacks(tick); main.postDelayed(tick, 500L)
    }

    fun setPlaying(playing: Boolean) {
        val p = player ?: return
        if (playing) p.play() else p.pause()
        session.setPlaybackState(playing, p.currentPosition)
    }
    fun seek(positionMs: Double) { player?.seekTo(positionMs.toLong()) }

    override fun onDetachedFromWindow() {
        main.removeCallbacks(tick)
        if (currentRefStr != null && lastEmittedFor != currentRefStr) emitMetrics("aborted")
        player?.release(); player = null
        focus.abandon()
        session.release()
        super.onDetachedFromWindow()
    }

    private fun emitMetrics(outcome: String) {
        val ref = currentRefStr ?: return
        if (lastEmittedFor == ref) return
        lastEmittedFor = ref

        val firstCanplayMs: Double? = if (loadStartedAt != 0L && canplayAt != 0L)
            (canplayAt - loadStartedAt).toDouble() else null
        val totalPlayMs: Double? = if (loadStartedAt != 0L)
            (System.currentTimeMillis() - loadStartedAt).toDouble() else null

        val src = currentSource as? cc.musicme.offline.StreamSource
        val m = src?.metrics()
        onMetrics(mapOf(
            "v" to 1,
            "trackRef" to ref,
            "outcome" to outcome,
            "bootstrapMs" to (m?.bootstrapMs as Any?),
            "firstKeyMs" to 0,
            "firstRangeMs" to (m?.firstRangeMs as Any?),
            "firstCanplayMs" to (firstCanplayMs as Any?),
            "totalPlayMs" to (totalPlayMs as Any?),
            "bufferUnderruns" to bufferUnderruns,
            "sessionRotations" to (m?.sessionRotations ?: 0),
            "fileSizeBytes" to (m?.fileSizeBytes as Any?),
        ))
    }
}

object PlayerConfig {
    @Volatile var workerUrl: String? = null
    @Volatile var tokenProvider: (suspend () -> String)? = null
    @Volatile var currentToken: String? = null
}

object RemoteEvents {
    @Volatile var onNext: (() -> Unit)? = null
    @Volatile var onPrev: (() -> Unit)? = null
}
